CI Episode Content (#20)
ImperialCollegeLondon/intermediate_grad_school_git_course: _episodes/l1-05-continuous_integration.md

## Explanation of CI

Continuous integration (CI) is a software development practice that ensures all
contributions to a code base meet a defined standard. This is enforced via computational
workflows that apply checks and tests to a code commit. Failure of these workflows to
complete successfully is indicated via the code hosting platform and can be used to

> block code from being merged into branches.

Reviewer: I'd mention an example or two in passing of a defined standard (e.g. whitespace formatting, unit tests etc.) to make it more concrete for students.

Author: Done

> Continuous integration (CI) is often paired with additional workflows that run after a

Reviewer: Repetition:
```suggestion
CI is often paired with additional workflows that run after a
```

so that `job-2` will only run if `job-1` finished successfully. Or you can provide
additional expressions to limit when a job should run e.g. only run a job for a
particular branch. You can also define a single job that is run multiple times with
different parameters using a matrix. This can be used, for instance, to test code on
multiple different operating systems with different versions of Python. See [GitHub
Docs: Using a matrix for your jobs][matrix] for more information.

[matrix]: https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs
[jobs]: https://docs.github.com/en/actions/using-jobs/using-jobs-in-a-workflow

```yaml
    steps:
      - uses: actions/checkout@v3
      - run: echo 'hello world'
```

Individual steps within a job define the actual work to be carried out. The workflow
above defines two steps that work in different ways. The first step has a `uses` entry
to indicate that it should use a pre-packaged action. This is a powerful feature of
GitHub Actions; individual job steps can be packaged and shared for use in workflows in
different repositories. Actions that have been packaged this way can be found in the

> [GitHub MarketPlace]. Here we're using version 3 of the `checkout` action which is

Reviewer 2: Do we need a link here for `GitHub MarketPlace`?

Author: It's defined below - L324

>     on:
>       - push
>     jobs:
>       markdownlint:
>         runs-on: ubuntu-latest
>         steps:
>           - uses: actions/checkout@v3
>           - name: markdownlint-cli
>             uses: nosborn/github-action-markdown-cli@v3.2.0
>             with:
>               files: .
>     ```
>
> 1. Stage and commit `ci.yml` then push the repository to GitHub.
> 1. Your first CI run should have been triggered! Quickly, go to your repository on
>    GitHub and select the `Actions` tab. You should see a workflow with a glowing amber
>    dot next to the commit messaged you provided. This means that the workflow is
>    running.
> 1. Click on the commit message. You now get a breakdown of the individual jobs within
>    your workflow. It's only one job in this case - `markdownlint` - click on it to see
>    its progress. You can see the individual steps, and the output that they produce as
>    they run.
> 1. Before long the workflow will complete but, alas, it should be a failure. Go back
>    to the front page of the repository by clicking the `Code` tab. You should see your
>    commit marked with a red cross to indicate that it failed the CI. You should also

> >    receive a notification (after a few minutes) via the email address associated with

Reviewer 2: Is this the default? I do not receive notifications because I disabled it long ago, but I mention it just incase someone else is in the same situation and is surprised at not receiving the email.

Author: It does seem to be the default. Not sure if it's worth a note here.

Reviewer 2: Maybe a small callout, just in case.

>    should succeed. If it doesn't then try modifying the files again.
> 1. Once the CI is passing, go back to the `Code` tab and you should see a nice green
>    tick next to your latest commit.
>
{: .challenge}

[markdownlint-cli]: https://github.com/igorshubovych/markdownlint-cli
[md041]: https://github.com/DavidAnson/markdownlint/blob/main/doc/Rules.md#md041---first-line-in-a-file-should-be-a-top-level-heading

### Using CI with Pull Requests

We've seen some basic usage of GitHub Actions but, so far, its only utility is adding a
tick or cross in the GitHub UI. That's good, but CI can be even more useful when
combined with pull requests.

Let's say we've created a new branch that we want to merge into main. If we create a
pull request but our CI is failing in the new branch, we'll see something like the
following:

![Failing CI]({{ site.baseurl }}/fig/pr_failing_ci.png "Panel from GitHub pull request
user interface showing failing CI checks"){:class="img-responsive"}

GitHub makes the failure of the CI pretty apparent but, by default, it will still allow
the PR to be merged. At this point the CI is a useful aid to peer review but we can

> take things further by implementing some policy in the form of a "branch projection

Reviewer 2: ```suggestion
take things further by implementing some policy in the form of a "branch protection
```

did not successfully pass through CI first**. Creating a cast iron guarantee that all
code that has been accepted into the `main` branch meets a certain standard is very
powerful.

Let's see how to create a branch protection rule and how this changes the behaviour of
PR's:

1. Go to the `Settings` tab and select `Branches` from the left-hand side.
1. Select `Add branch protection rule`.
1. Set `main` as the `Branch name pattern` and check the boxes for `Require a pull
   request before merging` and `Require status checks to pass before merging`.
1. In the extra options that appear beneath the status checks checkbox, check `Require
   branches to be up to date before merging`. Using the search bar, find and select the
   names of any CI jobs that must pass to allow merging.
1. Scroll down and press `Create`. GitHub may ask you to confirm your password.

Now a CI failure for a pull request looks like this:

![Failing CI]({{ site.baseurl }}/fig/pr_protected_failing_ci.png "Panel from GitHub pull
request user interface showing failing CI checks with merging blocked due to a branch
protection rule"){:class="img-responsive"}

> Now its much harder to get anything passed peer review that doesn't meet the required

Reviewer: ```suggestion
Now it's much harder to get anything past peer review that doesn't meet the required
```

you may need additional setup actions if you have specific requirements for different
versions.

The value of this kind of workflow is pretty straight forward. You can check that a
freshly checked out version of your can be successfully built. You can run similar
builds across a variety of operating systems and compilers to ensure broad
compatibility.

```yaml
on:
  - push
jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - uses: actions/actions-setup-cmake@v1
        with:
          cmake-version: "3.3"
      - run: |
          mkdir build
          cd build
          cmake ..
          make all
```

### Run Tests

Writing tests is an important best practice for software development. Even better is
incorporating tests into your CI so you know they pass in a newly checked-out repository
on another computer.

The below shows an example of runnitg the tests of a Python project:

```yaml
on:
  - push
jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - uses: actions/setup-python@v4
        with:
          python-version: "3.10"

>       - run: pip install -r requirements-dev.txt
>       - run: pytest .

Reviewer: Might it be worth using the multiline form of `run` here? I suppose you might also need to explain it to students in this case.

```suggestion
      - run: |
          pip install -r requirements-dev.txt
          pytest .
```

Author: Yeah tricky one. On balance I think probably best to avoid complicating things with additional yaml syntax even if that is how I'd probably write it IRL.

```

### Publish

Once we've accepted changes into our repository it can then be useful to trigger
actions. The below workflow builds and publishes a Docker image only when new commits

> are added to the `main` branch:

Reviewer: I think it would be worth briefly saying what docker is (maybe in brackets) and providing a link.

Author: Done :smiley:
